drivers/led/sn32/rgb_matrix_sn32f24xb.c

```diff
@@ -0,0 +1,426 @@
+}
+
+void shared_matrix_rgb_disable_pwm(void) {
+    // Disable PWM outputs on column pins
+    for(uint8_t y = 0; y < LED_MATRIX_COLS; y++) {
+        pwmDisableChannel(&PWMD1,chan_col_order[y]);
+    }
+}
+
+void shared_matrix_rgb_disable_leds(void) {
+    // Disable LED outputs on RGB channel pins
+    for (uint8_t x = 0; x < LED_MATRIX_ROWS_HW; x++) {
+        writePinLow(led_row_pins[x]);
+    }
+}
+
+void update_pwm_channels(PWMDriver *pwmp) {
+    matrix_row_t row_shifter = MATRIX_ROW_SHIFTER;
+    for(uint8_t col_idx = 0; col_idx < LED_MATRIX_COLS; col_idx++, row_shifter <<= 1) {
+        #if(DIODE_DIRECTION == ROW2COL)
+            // Scan the key matrix column
+          matrix_read_rows_on_col(current_matrix,col_idx,row_shifter);
+        #endif
+        uint8_t led_index = g_led_config.matrix_co[row_idx][col_idx];
+        // Check if we need to enable RGB output
+        if (led_state[led_index].b != 0) enable_pwm |= true;
+        if (led_state[led_index].g != 0) enable_pwm |= true;
+        if (led_state[led_index].r != 0) enable_pwm |= true;
+        // Update matching RGB channel PWM configuration
+        switch(current_row % LED_MATRIX_ROW_CHANNELS) {
+        case 0:
+                if(enable_pwm) pwmEnableChannelI(pwmp,chan_col_order[col_idx],led_state[led_index].b);
+            break;
+        case 1:
+                if(enable_pwm) pwmEnableChannelI(pwmp,chan_col_order[col_idx],led_state[led_index].g);
+            break;
+        case 2:
+                if(enable_pwm) pwmEnableChannelI(pwmp,chan_col_order[col_idx],led_state[led_index].r);
+            break;
+        default:
+            ;
+        }
+    }
+}
+void rgb_callback(PWMDriver *pwmp) {
+    // Disable the interrupt
+    pwmDisablePeriodicNotification(pwmp);
+    // Advance to the next LED RGB channels
+    current_row++;
+    if(current_row >= LED_MATRIX_ROWS_HW) current_row = 0;
+    // Advance to the next key matrix row
+    if(current_row % LED_MATRIX_ROW_CHANNELS == 2) row_idx++;
+    if(row_idx >= LED_MATRIX_ROWS) row_idx = 0;
+    chSysLockFromISR();
+    // Disable LED output before scanning the key matrix
+    shared_matrix_rgb_disable_leds();
+    shared_matrix_rgb_disable_pwm();
+    #ifdef MATRIX_NO_SCAN
+    #   if(DIODE_DIRECTION == COL2ROW)
+        // Scan the key matrix row
+        matrix_read_cols_on_row(current_matrix, row_idx);
+    #   endif
+    #endif
+    update_pwm_channels(pwmp);
+    if(enable_pwm) writePinHigh(led_row_pins[current_row]);
+    chSysUnlockFromISR();
+    #ifdef MATRIX_NO_SCAN
+    //scan done, update the matrix
+    bool changed = memcmp(raw_matrix, current_matrix, sizeof(current_matrix)) != 0;
+    if (changed) memcpy(raw_matrix, current_matrix, sizeof(current_matrix));
+    has_changed = changed;
+    #endif
+    // Advance the timer to just before the wrap-around, that will start a new PWM cycle
+    pwm_lld_change_counter(pwmp, 0xFFFF);
+    // Enable the interrupt
+    pwmEnablePeriodicNotification(pwmp);
```

Comment:
Again, this should be `pwmEnablePeriodicNotificationI()`, which would require moving `chSysUnlockFromISR()`.